File: packages/python_backend/app/utils/file_node_tree_utils.py
from typing import Dict, List, Optional, Union, Any
from pydantic import BaseModel, Field
import math

from app.schemas.project_schemas import ProjectFile

class FileNode(BaseModel):
    is_folder: bool = Field(..., alias='_folder')
    file: Optional[ProjectFile] = None
    children: Optional[Dict[str, 'FileNode']] = None

    model_config = {
        "populate_by_name": True
    }

FileNode.model_rebuild()


def format_token_count(content: Union[str, int]) -> str:
    count = estimate_token_count(content) if isinstance(content, str) else content
    if count >= 1000:
        return f"{(count / 1000):.2f}".rstrip('0').rstrip('.') + 'k'
    return str(count)

def estimate_token_count(text: str, chars_per_token: int = 4) -> int:
    if not text:
        return 0
    length = len(text)
    if length == 0 or chars_per_token <= 0:
        return 0
    return math.ceil(length / chars_per_token)

def count_total_files(root: Dict[str, FileNode]) -> int:
    count = 0
    
    def recurse(node: FileNode):
        nonlocal count
        if node.is_folder and node.children:
            for child_node in node.children.values():
                recurse(child_node)
        elif node.file:
            count += 1
            
    for node_val in root.values():
        recurse(node_val)
    return count

def collect_files(node: FileNode) -> List[int]:
    ids: List[int] = []
    if node.is_folder and node.children:
        for child_node in node.children.values():
            ids.extend(collect_files(child_node))
    elif node.file and node.file.id is not None:
        ids.append(node.file.id)
    return ids

def calculate_folder_tokens(folder_node: FileNode, selected_files: List[int]) -> Dict[str, int]:
    total_tokens = 0
    selected_tokens = 0

    if folder_node.is_folder and folder_node.children:
        for child in folder_node.children.values():
            if child.is_folder:
                child_folder_tokens = calculate_folder_tokens(child, selected_files)
                total_tokens += child_folder_tokens['total_tokens']
                selected_tokens += child_folder_tokens['selected_tokens']
            elif child.file and child.file.content:
                tokens = estimate_token_count(child.file.content)
                total_tokens += tokens
                if child.file.id is not None and child.file.id in selected_files:
                    selected_tokens += tokens
    return {"selected_tokens": selected_tokens, "total_tokens": total_tokens}

def are_all_folder_files_selected(folder_node: FileNode, selected_files: List[int]) -> bool:
    all_files_in_folder = collect_files(folder_node)
    if not all_files_in_folder:
        return False 
    return all(file_id in selected_files for file_id in all_files_in_folder)

def is_folder_partially_selected(folder_node: FileNode, selected_files: List[int]) -> bool:
    if not folder_node.is_folder:
        return False
    all_files_in_folder = collect_files(folder_node)
    if not all_files_in_folder:
        return False
    selected_count = sum(1 for file_id in all_files_in_folder if file_id in selected_files)
    return 0 < selected_count < len(all_files_in_folder)

def toggle_file(file_id: int, selected_files: List[int]) -> List[int]:
    if file_id in selected_files:
        return [id_ for id_ in selected_files if id_ != file_id]
    else:
        return selected_files + [file_id]

def toggle_folder(folder_node: FileNode, select: bool, selected_files: List[int]) -> List[int]:
    all_files_in_folder = collect_files(folder_node)
    current_selected_set = set(selected_files)
    if select:
        for file_id in all_files_in_folder:
            current_selected_set.add(file_id)
    else:
        for file_id in all_files_in_folder:
            current_selected_set.discard(file_id)
    return list(current_selected_set) 

File: packages/shared/src/utils/file-tree-utils/file-node-tree-utils.ts
import type { ProjectFile, ProjectFileMap } from '@promptliano/schemas'
import type { TsconfigCache } from './import-resolver'

export type FileNode = {
  _folder: boolean
  file?: ProjectFile
  children?: Record<number, FileNode>
}

// take in raw string or number which would be an already estimated token count
export function formatTokenCount(content: string | number) {
  const count = typeof content === 'number' ? content : estimateTokenCount(content)

  // Format the count
  let formattedCount: string
  if (count >= 1000) {
    formattedCount = (count / 1000).toFixed(2).replace(/\.?0+$/, '') + 'k'
  } else {
    formattedCount = count.toString()
  }

  return formattedCount
}

export function estimateTokenCount(text: string, charsPerToken: number = 4): number {
  const length = text.length
  if (length === 0 || charsPerToken <= 0) {
    return 0
  }
  return Math.ceil(length / charsPerToken)
}

export function countTotalFiles(root: Record<string, FileNode>): number {
  let count = 0
  function recurse(node: FileNode) {
    if (node._folder && node.children) {
      Object.values(node.children).forEach(recurse)
    } else if (node.file) {
      count++
    }
  }
  Object.values(root).forEach(recurse)
  return count
}

export function collectFiles(node: FileNode): number[] {
  let ids: number[] = []
  if (node._folder && node.children) {
    for (const key of Object.keys(node.children)) {
      // @ts-expect-error - key is a number
      ids = ids.concat(collectFiles(node.children[key]))
    }
  } else if (node.file?.id) {
    ids.push(node.file.id)
  }
  return ids
}

export function calculateFolderTokens(
  folderNode: FileNode,
  selectedFiles: number[]
): { selectedTokens: number; totalTokens: number } {
  let total = 0
  let selected = 0

  if (folderNode._folder && folderNode.children) {
    for (const child of Object.values(folderNode.children)) {
      if (child._folder) {
        const childTokens = calculateFolderTokens(child, selectedFiles)
        total += childTokens.totalTokens
        selected += childTokens.selectedTokens
      } else if (child.file?.content) {
        const tokens = estimateTokenCount(child.file.content)
        total += tokens
        if (child.file.id && selectedFiles.includes(child.file.id)) {
          selected += tokens
        }
      }
    }
  }
  return { selectedTokens: selected, totalTokens: total }
}

export function areAllFolderFilesSelected(folderNode: FileNode, selectedFiles: number[]): boolean {
  const allFiles = collectFiles(folderNode)
  return allFiles.length > 0 && allFiles.every((id) => selectedFiles.includes(id))
}

export function isFolderPartiallySelected(folderNode: FileNode, selectedFiles: number[]): boolean {
  if (!folderNode._folder) return false
  const allFiles = collectFiles(folderNode)
  const selectedCount = allFiles.filter((id) => selectedFiles.includes(id)).length
  return selectedCount > 0 && selectedCount < allFiles.length
}

export function toggleFile(
  fileId: number,
  selectedFiles: number[],
  resolveImports: boolean,
  fileMap: ProjectFileMap,
  getRecursiveImports: (fileId: number, allFiles: ProjectFile[], tsconfigCache: TsconfigCache) => number[],
  buildTsconfigAliasMap: (allFiles: ProjectFile[]) => TsconfigCache
): number[] {
  if (selectedFiles.includes(fileId)) {
    return selectedFiles.filter((id) => id !== fileId)
  } else {
    if (resolveImports) {
      const allFiles = Array.from(fileMap.values())
      const tsconfigCache = buildTsconfigAliasMap(allFiles)
      const dependencies = getRecursiveImports(fileId, allFiles, tsconfigCache)
      return [...selectedFiles, fileId, ...dependencies]
    }
    return [...selectedFiles, fileId]
  }
}

export function toggleFolder(folderNode: FileNode, select: boolean, selectedFiles: number[]): number[] {
  const allFiles = collectFiles(folderNode)
  const newSet = new Set<number>(selectedFiles)
  if (select) {
    for (const id of allFiles) newSet.add(id)
  } else {
    for (const id of allFiles) newSet.delete(id)
  }
  return Array.from(newSet)
}
